### PR TITLE
Fixing Inventories, and Trade Offers, attempt 2

### PR DESCRIPTION
**Pull request by bartico6**
This time all of the stuff I mentioned in the previous PR is fixed, and everything is nicely compressed into a "one-commit-per-file-changed" form.
Again, recapping what this does:
1. If received an invalid inventory - downloading will be retried a few times before continuing.
2. TradeOfferManager didn't nullcheck when fetching a single offer - fixed.
3. Offer -> TradeOffer conversion didn't include missing assets, making checking "which side's items went missing" impossible. Fixed as well.

### CODE REVIEW

**Comment by bartico6** [comment]
AppVeyor Y U FAIL ;____;

File: SteamTrade/Inventory.cs [hunk 1]
```diff
@@ -15,11 +15,19 @@ public class Inventory
...
+            InventoryResponse result = null;
+            while ((result == null || result.result == null || result.result.items == null) && attempts <= 5)
+            {
+                var url = "http://api.steampowered.com/IEconItems_440/GetPlayerItems/v0001/?key=" + apiKey + "&steamid=" + steamId;
+                string response = steamWeb.Fetch(url, "GET", null, false);
+                result = JsonConvert.DeserializeObject<InventoryResponse>(response);
+                attempts++;
+            }
+            if (result == null || result.result == null || result.result.items == null)
+                throw Exception("Cannot fetch inventory of user #" + steamId);
```
Comment (nicolas-martin) [inline SteamTrade/Inventory.cs]: I think you forgot to add "new" when throwing an exception. Would explain the AppVeyor error (can't confirm right now though)

Comment (bartico6) [inline SteamTrade/Inventory.cs]: AppVeyor is failing right after installing NUnit so I don't think it's this line's fault, I'll correct the mistake though.

Comment (scholtzm) [inline SteamTrade/Inventory.cs]: You are right. It started right after this exception was added: https://ci.appveyor.com/project/Jessecar96/steambot/build/1.0.72

**Comment by BlueRaja** [comment]
Could you please test this to make sure it actually works?  The fact that it didn't even compile until recently is a pretty good indication that you did not test it yet...

**Comment by nicolas-martin** [comment]
I'm not sure I understand the need for an 'isMissing" parameter. What case would you want to create and asset where that is true?

File: SteamTrade/TradeOffer/TradeOffer.cs [hunk 2]
```diff
@@ -532,22 +517,27 @@ public class TradeAsset : IEquatable<TradeAsset>
                 [JsonProperty("currencyid"), JsonConverter(typeof(ValueStringConverter))]
                 public long CurrencyId { get; set; }
 
-                public void CreateItemAsset(long appId, long contextId, long assetId, long amount)
+                [JsonProperty("missing")]
+                public bool IsMissing { get; set; }
+
+                public void CreateItemAsset(long appId, long contextId, long assetId, long amount, bool missing)
```
Comment (scholtzm) [inline SteamTrade/TradeOffer/TradeOffer.cs]: `missing` could/should have default value set to `false` I suppose?

File: SteamTrade/TradeOffer/TradeOffer.cs [hunk 3]
```diff
@@ -532,22 +517,27 @@ public class TradeAsset : IEquatable<TradeAsset>
...
                     this.AppId = appId;
                     this.ContextId = contextId;
                     this.AssetId = assetId;
                     this.Amount = amount;
                     this.CurrencyId = 0;
+                    this.IsMissing = missing;
                 }
 
-                public void CreateCurrencyAsset(long appId, long contextId, long currencyId, long amount)
+                public void CreateCurrencyAsset(long appId, long contextId, long currencyId, long amount, bool missing)
```
Comment (scholtzm) [inline SteamTrade/TradeOffer/TradeOffer.cs]: same as above

**Comment by scholtzm** [comment]
btw. @BlueRaja currently TradeOfferManager is private in the Bot class. This is quite impractical since user often needs to fetch trade offer manually. Maybe this change could be added in this PR.

**Comment by ThereWasADream** [comment]
This could be wrong as I have only been teaching myself C# since about March, but...regarding ONLY the Inventory.cs while loop (I did not use the other parts of the pull request):

Without this, inventory requests seem more vulnerable to 503's from Steam (which happen a lot!)

With this, other WebRequest try-catch blocks in other Steambot documents and RetryWebRequest  appear to either be broken, or their error logs are overridden.

Hope this makes sense...

**Comment by ThereWasADream** [comment]
I rescind my earlier comment. I did not realize I was not properly handling the exception bartico wrote. Once I did so, the changes to inventory.cs were very helpful. I still have not tested the other parts of this pull request.

**Comment by bartico6** [comment]
The PR seems pretty much dead but feel free to use the changes - maybe someone will feel like reviving this and making a PR that will implement those - I'm fine with that ^^

**Comment by BlueRaja** [comment]
I merged your other PR instead, as this PR adds other complications that I and others don't see the value of.

**Comment by Jessecar96** [comment]
Closing as this is no longer compatible. Make a new PR if you want to add this again.

**Pull request closed without merging**